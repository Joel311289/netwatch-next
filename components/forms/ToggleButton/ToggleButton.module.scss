$button-border-radius: 8px;
$button-border-width: 1px;
$button-border: $button-border-width solid var(--color-border);

@mixin radius($key: 'left') {
  @if $key == 'left' {
    border-top-left-radius: $button-border-radius;
    border-bottom-left-radius: $button-border-radius;
  } @else {
    border-top-right-radius: $button-border-radius;
    border-bottom-right-radius: $button-border-radius;
  }
}

/** ToggleButton (main component) **/

.toggleButtonWrapper {
  height: 48px;
  width: fit-content;
}

/** ToggleButtonItem (child component) **/

.toggleButtonItemWrapper {
  @include font($size: 16px, $color: var(--color-text));

  background-color: transparent;
  cursor: pointer;
  border: $button-border;
  padding: 0 15px;
  height: 100%;
  transition: all 0.2s;
  margin-left: calc(-1 * #{$button-border-width});

  &:hover {
    background-color: rgb(var(--rgb-background-color-clear-hover) / 8%);
  }

  &.selected {
    border-color: var(--color-primary-dark) !important;
    background-color: rgb(var(--rgb-background-color-primary-hover) / 70%);
    z-index: 2;
  }

  &:first-child {
    @include radius('left');
  }

  &:last-child {
    @include radius('right');
  }
}
