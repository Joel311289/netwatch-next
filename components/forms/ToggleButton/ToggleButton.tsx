import { createContext, useCallback, useContext, useEffect, useRef, useState } from 'react';
import cn from 'classnames';
import { ElementChildren, ElementHTML, FunctionVoid } from '@/types';
import { withChildrenFiltered } from '@/hoc/withChildrenFiltered';
import { Space } from '@/components/layout';
import styles from '@/components/forms/ToggleButton/ToggleButton.module.scss';

/* -------------------------------------------------------------------------- */
/** Context **/
/* -------------------------------------------------------------------------- */

type ToggleButtonContextProps = typeof defaultValue & {
  activeButton?: string;
  onChangeButton?: FunctionVoid<string>;
};

const defaultValue = {
  activeButton: ''
};

const ToggleButtonContext = createContext<ToggleButtonContextProps>(defaultValue);

const useToggleButtonContext = () => useContext(ToggleButtonContext);

/* -------------------------------------------------------------------------- */
/** ToggleButtonItem (child component) **/
/* -------------------------------------------------------------------------- */

type ToggleButtonItemProps = ElementHTML &
  ElementChildren & {
    value: string;
    selected?: boolean;
  };

const ToggleButtonItem = ({ children, className, value, selected }: ToggleButtonItemProps) => {
  const { activeButton, onChangeButton } = useToggleButtonContext();
  const changeRef = useRef<FunctionVoid<string> | undefined>(onChangeButton);

  const changeButton = useCallback(
    () => changeRef.current && (changeRef.current as FunctionVoid<string>)(value),
    [value]
  );

  useEffect(() => {
    selected && changeButton();
  }, [changeButton, selected]);

  return (
    <button
      aria-label="button"
      aria-pressed={activeButton === value}
      className={cn(
        styles.toggleButtonItemWrapper,
        activeButton === value && styles.selected,
        className
      )}
      onClick={changeButton}>
      {children}
    </button>
  );
};

/* -------------------------------------------------------------------------- */
/** ToggleButton (main component) **/
/* -------------------------------------------------------------------------- */

export type ToggleButtonProps = typeof defaultProps &
  ElementHTML &
  ElementChildren<Array<JSX.Element>> & {
    onChange?: FunctionVoid<string>;
  };

const defaultProps = {};

const ToggleButton = ({ className, children, onChange }: ToggleButtonProps) => {
  const [activeButton, setActiveButton] = useState<string>(
    children && children.length ? children[0].props.value : ''
  );

  const handleChangeButton = useCallback(
    (key: string) => {
      setActiveButton(key);
      onChange && onChange(key);
    },
    [onChange]
  );

  return (
    <ToggleButtonContext.Provider value={{ activeButton, onChangeButton: handleChangeButton }}>
      <Space align="center" className={cn(styles.toggleButtonWrapper, className)}>
        {children}
      </Space>
    </ToggleButtonContext.Provider>
  );
};

ToggleButton.defaultProps = defaultProps;

const ToggleButtonWithChildrenFiltered = withChildrenFiltered(ToggleButton, {
  Item: ToggleButtonItem
});

export default ToggleButtonWithChildrenFiltered;
