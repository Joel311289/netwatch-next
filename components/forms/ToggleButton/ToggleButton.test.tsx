import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ToggleButton from '@/components/forms/ToggleButton/ToggleButton';

const items = [
  { label: 'Item 1', value: 'item1', selected: false },
  { label: 'Item 2', value: 'item2' }
];

const ToggleButtonItems = () => {
  return items.map(({ label, value, selected }) => (
    <ToggleButton.Item key={value} value={value} selected={selected}>
      {label}
    </ToggleButton.Item>
  ));
};

describe('Tests ToggleButton component', () => {
  test('renders component correctly', () => {
    const { container } = render(<ToggleButton>{ToggleButtonItems()}</ToggleButton>);

    expect(container).toMatchSnapshot();
  });

  test('renders children correctly', () => {
    render(<ToggleButton>{ToggleButtonItems()}</ToggleButton>);

    expect(screen.getAllByRole('button').length).toBe(items.length);
    expect(screen.getAllByText(/Item [1,2]/i).length).toBe(items.length);
  });

  test('renders children with child selected', () => {
    items[0].selected = true;

    render(<ToggleButton>{ToggleButtonItems()}</ToggleButton>);

    expect(screen.getByText(items[0].label)).toHaveAttribute('aria-pressed', 'true');
  });

  test('calls onChange prop and selects child', () => {
    const { label, value } = items[0];
    const handleChange = jest.fn();

    render(<ToggleButton onChange={handleChange}>{ToggleButtonItems()}</ToggleButton>);

    userEvent.click(screen.getByText(label));

    expect(handleChange).toHaveBeenCalledWith(value);
    expect(screen.getByText(label)).toHaveAttribute('aria-pressed', 'true');
  });
});
