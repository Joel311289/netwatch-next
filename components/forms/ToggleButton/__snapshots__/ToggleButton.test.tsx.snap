// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests ToggleButton component renders component correctly 1`] = `
<div>
  <div
    class="wrapper wrapper"
  >
    <button
      aria-pressed="false"
      class="itemWrapper"
    >
      Item 1
    </button>
    <button
      aria-pressed="false"
      class="itemWrapper"
    >
      Item 2
    </button>
    <style>
      
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          
          
          align-items: center
        }
      
    </style>
  </div>
</div>
`;
